zone/spells.cpp

```diff
@@ -3211,29 +3205,55 @@ int Mob::CheckStackConflict(uint16 spellid1, int caster_level1, uint16 spellid2,
 // spells 1-50: no restrictons
 // 51-65: SpellLevel/2+15
 // 66+ Group Spells 62, Single Target 61
-bool Mob::CheckSpellLevelRestriction(uint16 spell_id)
+bool Mob::CheckSpellLevelRestriction(Mob *caster, uint16 spell_id)
 {
-	return true;
-}
+	bool check_for_restrictions = false;
+	bool can_cast = true;
 
-bool Client::CheckSpellLevelRestriction(uint16 spell_id)
-{
-	int SpellLevel = GetMinLevel(spell_id);
+	if (!caster) {
+		LogSpells("CheckSpellLevelRestriction: No caster");
+		return false;
+	}
 
-	// Only check for beneficial buffs
-	if (IsBuffSpell(spell_id) && IsBeneficialSpell(spell_id)) {
-		if (SpellLevel > 65) {
-			if (IsGroupSpell(spell_id) && GetLevel() < 62)
-				return false;
-			else if (GetLevel() < 61)
-				return false;
-		} else if (SpellLevel > 50) { // 51-65
-			if (GetLevel() < (SpellLevel / 2 + 15))
-				return false;
+	// NON GM clients might be restricted by rule setting
+	if (caster->IsClient()) {
+		if (RuleB(Spells, BuffLevelRestrictions) && !caster->CastToClient()->GetGM()) {
```

Comment:
We should emit a log when this logic checks GM

So many times people waste hours troubleshooting a non-issue when they realize an effect wasn't working because they just had the GM flag on.

Something in `LogSpells` at the very least

Reply:
If I add this, it will be a LogMessage that comes out every time a GM casts a spell when RuleB(Spells, BuffLevelRestrictions is set, because the code does not restrict GMS regardless of the rule.  Do you want that?

Reply:
Doesn't hurt if its in a categorized log (Spells)

Reply:
Done.